src/error.rs

```diff
@@ -1,7 +1,52 @@
 //! Contains error types and other error handling tools.
 
+use core::fmt;
+
+use crate::consensus::encode;
+use crate::internal_macros::write_err;
 pub use crate::parse::ParseIntError;
 
+/// A general error code, other errors should implement conversions to/from this
+/// if appropriate.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum Error {
+    /// Encoding error
+    Encode(encode::Error),
+    /// The header hash is not below the target
+    BlockBadProofOfWork,
+    /// The `target` field of a block header did not match the expected difficulty
+    BlockBadTarget,
+}
```

Comment:
WTF, I never noticed until now that this has so few variants. I'd expect this to contain all error types in the whole library. Not for this PR OFC.

Reply:
Heh, historically we've used `util::Error` as "the default error type" unless we had to add a bunch of variants at once (e.g. in PSBT) in which case we added our own error type.

I think a long time ago I had an intention to add `From` impls for all the other types to `util::Error` but I never really followed up on that.